### PR TITLE
Updated to connectionless APIs (fix #105)

### LINKED ISSUE #105: Using FusedLocationProviderClient?
Question: Are you using FusedLocationProviderClient? Cos FusedLocationProviderApi is already deprecated. Thanks

### PR DESCRIPTION
**Pull request by victor-denisenko**
Fix #105 

Remove deprecated FusedLocationProviderApi, use [FusedLocationProviderClient](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient).

**Public API changes**:
- Delete fields: 
 GooglePlayServicesConfiguration.java : delete failOnConnectionSuspended, suspendedConnectionRetryCount fields and get/set methods.
 FailType.java : delete GOOGLE_PLAY_SERVICES_CONNECTION_FAIL field.
 These fields are unused, new api is connectionless.

Other changes:
- Defaults.java : delete FAIL_ON_CONNECTION_SUSPENDED, SUSPENDED_CONNECTION_RETRY_COUNT
- add some tests
- update README

**About deleting public api fields**: I don't known, maybe save this fields and mark them as deprecated? @yayaa, what your opinion?

### CODE REVIEW

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProvider.java [hunk 1]
```diff
@@ -192,9 +214,14 @@ void continueWithDefaultProviders() {
                 getListener().onLocationFailed(FailType.GOOGLE_PLAY_SERVICES_NOT_AVAILABLE);
             }
         } else {
-            LogUtils.logI("Attempting to get location from default providers...");
-            setLocationProvider(getSourceProvider().createDefaultLocationProvider());
-            activeProvider.get();
+            // check if activeProvider is already DefaultLocationProvider
+            if (activeProvider instanceof DefaultLocationProvider) {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProvider.java:218]: Any reason for adding this check? Also, what is the expected behaviour if it is already default provider?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 2]
```diff
@@ -26,8 +32,12 @@
     private final WeakReference<FallbackListener> fallbackListener;
 
     private boolean settingsDialogIsOn = false;
-    private boolean waitingForConnectionToRequestLocationUpdate = true;
-    private int suspendedConnectionIteration = 0;
+
+    /**
+     * Tracks the status of the location updates request.
+     */
+    boolean mRequestingLocationUpdates = false;
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:39]: Please follow the general naming convention in the library, no Hungarian Notation.

Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:39]: Also, you don't really need this. Please use `setWaiting` instead, since it is used from multiple parts in the provider as well.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 3]
```diff
@@ -98,6 +105,7 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                 LogUtils.logI("We got settings changed, requesting location update...");
                 requestLocationUpdate();
             } else {
+                mRequestingLocationUpdates = false;
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:108]: You haven't requested location updates yet, you shouldn't set to false here

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 4]
```diff
@@ -106,94 +114,97 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
     }
 
     @Override
-    public void onConnected(Bundle bundle) {
-        LogUtils.logI("GoogleApiClient is connected.");
-        boolean locationIsAlreadyAvailable = false;
+    public void onConnected() {
+        LogUtils.logI("Start request location updates.");
 
-        if (getConfiguration().googlePlayServicesConfiguration().ignoreLastKnowLocation()) {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:113]: I don't see how this is related to any of the other changes? Please keep the `lastKnowLocation` logic preserved here.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 5]
```diff
@@ -106,94 +114,97 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
     }
 
     @Override
-    public void onConnected(Bundle bundle) {
-        LogUtils.logI("GoogleApiClient is connected.");
-        boolean locationIsAlreadyAvailable = false;
+    public void onConnected() {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:117]: Too many logic change here, the only change you'd need is as following

```
if (getConfiguration().keepTracking() || !locationIsAlreadyAvailable
              || waitingForConnectionToRequestLocationUpdate) {
            waitingForConnectionToRequestLocationUpdate(false);
            locationRequired();
```

to

```
if (getConfiguration().keepTracking() || !locationIsAlreadyAvailable) {
            locationRequired();
```

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 6]
```diff
@@ -106,94 +114,97 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:
+                    // Location settings are not satisfied. But could be fixed by showing the user
+                    // a dialog.
+                    try {
+                        // Cast to a resolvable exception.
+                        resolveSettingsApi((ResolvableApiException) exception);
+
+                        break;
+                    } catch (ClassCastException e) {
+                        // Ignore, should be an impossible error.
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:186]: Please do not `try/catch` and ignore it. 
If casting here is safe, then remove try/catch - being aggressive here would make more sense. 

On the other hand, can we not call `resultTask.getResult(ResolvableApiException.class)` directly?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 7]
```diff
@@ -205,20 +216,72 @@ void resolveSettingsApi(Status status) {
...
-            }
+    void checkLastKnowLocation() {
+        getSourceProvider().getLocationAvailability()
+                .addOnSuccessListener(new OnSuccessListener<LocationAvailability>() {
+                    /**
+                     * Returns the availability of location data. When isLocationAvailable() returns true, then the location returned by getLastLocation() will be reasonably up to date within the hints specified by the active LocationRequests.
+                     *
+                     * If the client isn't connected to Google Play services and the request times out, null is returned.
+                     *
+                     * Note it's always possible for getLastLocation() to return null even when this method returns true (e.g. location settings were disabled between calls).
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:227]: If so, does it even worth to call `availability` in the first place? It seems like calling `getLastLocation` directly would be sufficient, considering the fact that on its absence it will trigger location request anyway.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 8]
```diff
@@ -205,20 +216,72 @@ void resolveSettingsApi(Status status) {
...
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        LogUtils.logI("LastKnowLocation is not available.");
+
+                        requestLocation(false);
+                    }
+                });
+    }
+
+    void requestLocation(boolean locationIsAlreadyAvailable) {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:279]: Please avoid introducing a new method to request a location update, use either `locationRequired` or `requestLocationUpdate` depending on the conditions.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 9]
```diff
@@ -273,21 +327,32 @@ void failed(@FailType int type) {
...
-        waitingForConnectionToRequestLocationUpdate = isWaiting;
-    }
-
     // For test purposes
     void setDispatcherLocationSource(GooglePlayServicesLocationSource googlePlayServicesLocationSource) {
         this.googlePlayServicesLocationSource = googlePlayServicesLocationSource;
     }
 
-    private GooglePlayServicesLocationSource getSourceProvider() {
+    GooglePlayServicesLocationSource getSourceProvider() {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:335]: There is no reason for this to be not-private. Please revert this change

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java [hunk 10]
```diff
@@ -4,133 +4,94 @@
...
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
 import com.yayandroid.locationmanager.constants.RequestCode;
 
-class GooglePlayServicesLocationSource implements LocationListener, ConnectionCallbacks, OnConnectionFailedListener,
-      ResultCallback<LocationSettingsResult> {
+class GooglePlayServicesLocationSource extends LocationCallback {
 
-    private final GoogleApiClient googleApiClient;
+    private final FusedLocationProviderClient googleApiClient;
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:26]: rename the variable as well?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java [hunk 11]
```diff
@@ -4,133 +4,94 @@
...
+    Task<Location> getLastLocation() {
+        return googleApiClient.getLastLocation();
     }
 
     @Override
-    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
-        if (sourceListener != null) sourceListener.onConnectionFailed(connectionResult);
-    }
+    public void onLocationResult(LocationResult locationResult) {
+        if (locationResult == null) {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:88]: This logic belongs to provider, please just redirect here.

File: library/src/test/java/com/google/android/gms/location/Helper.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package com.google.android.gms.location;
+
+import androidx.annotation.NonNull;
+
+public class Helper {
```
Comment (yayaa) [inline library/src/test/java/com/google/android/gms/location/Helper.java:5]: It is only used in one test class, please move it there as private method instead of having separate helper class for it

File: library/build.gradle [hunk 13]
```diff
@@ -24,7 +24,7 @@ android {
 
 dependencies {
     implementation libraries.appCompat
-    implementation libraries.googlePlayServices
+    api libraries.googlePlayServices
```
Comment (yayaa) [inline library/build.gradle:27]: This change should not be required, could you please revert back to `implementation`?

**Comment by yayaa** [comment]
Hey @victor-denisenko, thanks a lot for the PR. 
Specially for taking time to integrating the new API, updating tests and ReadMe. 
Really appreciated. 🙏 

Unfortunately, it contains a lot of different changes. I cannot accept it as is.
Would you mind splitting into several PRs as following? 

- Deprecating onStatusChanged
- GooglePlayService Dialog cancel / dismiss handling
- Adding leakCanary
- Replacing FusedLocationProviderApi with FusedLocationProviderClient (for this one, I have tried to provide feedback as much as possible already, please keep them in consideration while creating separate PRs.)
--------
- Library version update, I'll take care of this. Since it won't release automatically and I need to do manual operation anyway. (Self note: Cmon dude! Update this release process already.)

File: library/build.gradle (hunk 13 above)
Comment (victor-denisenko) [reply library/build.gradle:27]: In this way user cannot call LocationRequest.create() in LocationConfiguration.Builder#locationRequest(LocationRequest) in app, because googlePlayServices dependency is hidden.  
But yes, user can add googlePlayServices dependency in own app.
Return anyway to `implementation`?

Comment (yayaa) [reply library/build.gradle:27]: Yes, that is intentional. They only need to create `LocationRequest` if they want to use GooglePlayServices, which in that case they should have it included explicitly. Otherwise it is not necessary, making it `api` instead of `implementation` will prevent using this library without GooglePlayServices, which is not what we want.
 
So, please do revert it to `implementation`.

**Comment by victor-denisenko** [comment]
Hi, 
Thanks for the review, I will try to check code changes additionally and write answers soon.
Yes, I agree with splitting changes into several PRs.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 6 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java]: Split listener to onSuccess/onFailure, remove try/catch.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 7 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java]: Delete getLocationAvailability() call, use getLastLocation() directly

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java (hunk 10 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java]: Renamed to fusedLocationProviderClient

File: library/src/test/java/com/google/android/gms/location/Helper.java (hunk 12 above)
Comment (victor-denisenko) [reply library/src/test/java/com/google/android/gms/location/Helper.java]: Tests updated, class deleted

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 5 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:107]: This code moved to requestLocation(), because checkLastKnowLocation() is task now, we cannot get locationIsAlreadyAvailable value in this function.

Same logic is saved, I think.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 8 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:228]: This needed, because checkLastKnowLocation() is task now. Function stores some code from old onConnected(Bundle).

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java (hunk 11 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java]: You mean call `sourceListener.onLocationChanged(locationResult)` without loop in source?

Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java]: I mean, create `onLocationResult` method in provider and redirect this one to there, so that logic can be handled there and tested. This class should only be responsible to provide instances and redirect callbacks, so no test is needed. 

```
sourceListener.onLocationResult(locationResult);
```

and then, in that method, have this loop and eventually call `onLocationChanged` from it.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProvider.java [hunk 14]
```diff
@@ -155,14 +155,14 @@ void askForGooglePlayServices(int gpServicesAvailability) {
...
         gpServicesDialog = getSourceProvider().getGoogleApiErrorDialog(getActivity(), gpServicesAvailability,
-              RequestCode.GOOGLE_PLAY_SERVICES, new DialogInterface.OnCancelListener() {
-                  @Override
-                  public void onCancel(DialogInterface dialog) {
-                      LogUtils.logI("GooglePlayServices error could've been resolved, "
-                            + "but user canceled it.");
-                      continueWithDefaultProviders();
-                  }
-              });
+                RequestCode.GOOGLE_PLAY_SERVICES, new DialogInterface.OnCancelListener() {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationProvider.java:158]: Can you please clean up these spaces since there is no actual change here this file should not be in this PR?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationSource.java [hunk 15]
```diff
@@ -44,7 +44,7 @@ boolean isGoogleApiErrorUserResolvable(int gpServicesAvailability) {
     }
 
     @Nullable Dialog getGoogleApiErrorDialog(Activity activity, int gpServicesAvailability, int requestCode,
-          OnCancelListener onCancelListener) {
+                                             OnCancelListener onCancelListener) {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/DispatcherLocationSource.java:47]: Can you please clean up these spaces since there is no actual change here this file should not be in this PR?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 16]
```diff
@@ -290,4 +302,19 @@ private GooglePlayServicesLocationSource getSourceProvider() {
...
+    private void removeLocationUpdates() {
+        LogUtils.logI("Stop location updates...");
+
+        // not getSourceProvider, because we don't want to create if it doesn't already exist
+        if (googlePlayServicesLocationSource != null) {
+            googlePlayServicesLocationSource.removeLocationUpdates()
+                    .addOnCompleteListener(new OnCompleteListener<Void>() {
+                        @Override
+                        public void onComplete(@NonNull Task<Void> task) {
+                            setWaiting(false);
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:314]: This might cause leak. 
Please just do following instead
```
setWaiting(false);
googlePlayServicesLocationSource.removeLocationUpdates();
```

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 8 above)
Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:241]: I see

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 17]
```diff
@@ -106,94 +106,95 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+        requestLocationUpdate();
+    }
+
+    @Override
+    public void onFailure(@NonNull Exception exception) {
+        int statusCode = ((ApiException) exception).getStatusCode();
+
+        switch (statusCode) {
             case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:
+                setWaiting(false);
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:166]: You don't need this, any `...Fail` eventually does that under the hood.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 18]
```diff
@@ -106,94 +106,95 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
                 // Location settings are not satisfied. But could be fixed by showing the user
                 // a dialog.
-                resolveSettingsApi(status);
+                // Cast to a resolvable exception.
+                resolveSettingsApi((ResolvableApiException) exception);
+
+                break;
+            default:
+                // for other CommonStatusCodes values
+                setWaiting(false);
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:182]: same as above

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 19]
```diff
@@ -237,14 +262,9 @@ void requestLocationUpdate() {
...
-        if (getSourceProvider().isGoogleApiClientConnected()) {
-            LogUtils.logI("Requesting location update...");
-            getSourceProvider().requestLocationUpdate();
-        } else {
-            LogUtils.logI("Tried to requestLocationUpdate, but GoogleApiClient wasn't connected. Trying to connect...");
-            waitingForConnectionToRequestLocationUpdate(true);
-            getSourceProvider().connectGoogleApiClient();
-        }
+        LogUtils.logI("Requesting location update...");
+        setWaiting(true);
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:266]: This is set on `get` already, you don't need to set it again. 
Because get is the trigger point, and once this is true - there is either `fail` or `onLocationChanged` calls can happen, which will set to false - and both are terminal. There is no need to set it to true, it will misbehave on the case `keepTracking` 

So, please remove it :)

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java [hunk 20]
```diff
@@ -106,94 +106,95 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
-            case LocationSettingsStatusCodes.SUCCESS:
-                // All location settings are satisfied. The client can initialize location
-                // requests here.
-                LogUtils.logI("We got GPS, Wifi and/or Cell network providers enabled enough "
-                      + "to receive location as we needed. Requesting location update...");
-                requestLocationUpdate();
-                break;
+    public void onLocationResult(@Nullable LocationResult locationResult) {
+        if (locationResult == null) {
+            // Do nothing, wait for other result
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:142]: Is that a thing? Are you sure we shouldn't be failing here?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java [hunk 21]
```diff
@@ -4,133 +4,91 @@
...
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception exception) {
+                        if (sourceListener != null) sourceListener.onFailure(exception);
+                    }
+                });
+    }
+
+    void startSettingsApiResolutionForResult(@NonNull ResolvableApiException resolvable, Activity activity) throws SendIntentException {
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:68]: Let's follow the Android structure here, please have the `Activity` as first parameter in the method

Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:68]: Sorry just noticed that's how it was originally, so you may ignore this comment if you like to.
Have no idea why I have done that way in the first place 🤦

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java [hunk 22]
```diff
@@ -4,133 +4,91 @@
...
+        resolvable.startResolutionForResult(activity, RequestCode.SETTINGS_API);
     }
 
     @SuppressWarnings("ResourceType")
```
Comment (yayaa) [inline library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:72]: Is this still needed?

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+    public Task<TResult> addOnCanceledListener(@NonNull Activity var1, @NonNull OnCanceledListener var2) {
+        if (isCanceled()) {
+            var2.onCanceled();
+        }
+
+        return this;
+    }
+
+    public void setResult(@Nullable TResult var1) {
+        this.result = var1;
```
Comment (yayaa) [inline library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java:182]: Can you call `success` and `completed` here?
So that you don't need to override `isSuccessful` and `isComplete`

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    public boolean trySetResult(TResult var1) {
+        this.result = var1;
+
+        return true;
+    }
+
+    public void setException(@NonNull Exception var1) {
```
Comment (yayaa) [inline library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java:191]: Same here, to call fail or cancel (whatever the API allows you)
so that you dont need to override `isCanceled`

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 20 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:142]: As I can understand in obfuscated code, `locationResult` can be null if the first number in Parcel var0 header is 0:
```
// Called from here com.google.android.gms.location.zzv.dispatchTransaction()
// Class com.google.android.gms.internal.location.zzc
public static <T extends Parcelable> T zza(Parcel var0, Creator<T> var1) {
        return var0.readInt() == 0 ? null : (Parcelable)var1.createFromParcel(var0);
 }
```
I did'n find cause when first Int can be 0. Maybe in service disconnection phase all received LocationResults sets as unwanted?

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java (hunk 22 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:72]: Yes, for example if location is disabled (Go to Settings > Location > Use location > Turn off) this resolves issue.

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java (hunk 23 above)
Comment (victor-denisenko) [reply library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java]: No, listener not stored in class. MockSimpleTask updated, unused methods removed.

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java (hunk 24 above)
Comment (victor-denisenko) [reply library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java]: No, listener not stored in class. MockSimpleTask updated, unused methods removed.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 20 above)
Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:142]: Yes, that is fine, if they think API is nullable, then we should treat that way for sure.
My question was about "wait for other result" is that a thing? Is this a continuous task that calls the listener multiple times or will it terminate once it is called? If it will terminate, we need to terminate here as well, and potentially with an error.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java (hunk 22 above)
Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:72]: I meant `SuppressWarnings("ResourceType")` annotation :)

File: library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProviderTest.java [hunk 25]
```diff
@@ -62,74 +81,72 @@ public void setUp() throws Exception {
...
         googlePlayServicesLocationProvider.onResume();
 
-        verify(mockedSource, never()).clearGoogleApiClient();
+        verify(mockedSource, never()).requestLocationUpdate();
     }
 
     @Test
-    public void onResumeShouldConnectWhenLocationIsNotYetProvided() {
+    public void onResumeShouldRequestLocationUpdateWhenLocationIsNotYetProvided() {
+        when(mockedSource.getLastLocation()).thenReturn(new MockSimpleTask<>(((Location) null)));
```
Comment (yayaa) [inline library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProviderTest.java:103]: This is not clear what you expect Task to do. See my comment on `MockSimpleTask`

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.RuntimeExecutionException;
+import com.google.android.gms.tasks.Task;
+
+import java.util.concurrent.Executor;
+
+public class MockSimpleTask<TResult> extends Task<TResult> {
```
Comment (yayaa) [inline library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java:16]: Consider implementing this to reflect what you expect it to do, as in
`new MockSimpleTask<>(((Location) null))` is not clear what is your intention, obviously I know that you mean it to be successful now since I check the implementation, but you can build the API as following
```
MockSimpleTask mockSimpleTask = new MockSimpleTask<Location>();
mockSimpleTask.success(resultToVerify);
// verify expected action
verify(mockedSource).requestLocationUpdate();
```

Comment (yayaa) [reply library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java:16]: ...and to achieve this, you might need to keep list of listeners and call them when requested.

```
public class MockSimpleTask<TResult> extends Task<TResult> {

    private final List<OnSuccessListener<? super TResult>> onSuccessListeners = new ArrayList<>();

    public void success(TResult result) {
         for(OnSuccessListener<? super TResult> onSuccessListener: onSuccessListeners) {
            onSuccessListener.onSuccess(var1);
        }
    }

    @NonNull
    public Task<TResult> addOnSuccessListener(@NonNull OnSuccessListener<? super TResult> var1) {
        onSuccessListeners.add(var1);
        return this;
    }

}
```

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java (hunk 22 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java:72]: Yes. The fusedLocationProviderClient.requestLocationUpdates() needs any of ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION permission.

File: library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java (hunk 20 above)
Comment (victor-denisenko) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:142]: This is a continuous task, the connection will be active until removeLocationUpdates() is called.

Comment (yayaa) [reply library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java:142]: I see, thanks for the explanation.

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java (hunk 26 above)
Comment (victor-denisenko) [reply library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java:18]: Done

File: library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProviderTest.java (hunk 25 above)
Comment (victor-denisenko) [reply library/src/test/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProviderTest.java]: Done

File: library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java (hunk 26 above)
Comment (yayaa) [reply library/src/test/java/com/yayandroid/locationmanager/mocks/MockSimpleTask.java:18]: Thank you for the changes
